### PR TITLE
core(complex) make `conj` `constexpr`

### PR DESCRIPTION
**Pull request by romintomasetti**
This PR adds a missing `constexpr` for `Kokkos::conj`, and for the `real` and `imag` free functions.

Quick reference:
* https://en.cppreference.com/w/cpp/numeric/complex/conj

### CODE REVIEW

File: core/unit_test/TestComplex.hpp [hunk 1]
```diff
@@ -456,14 +456,13 @@ TEST(TEST_CATEGORY, complex_operations_arithmetic_types_overloads) {
...
-  ASSERT_EQ(Kokkos::conj(3.), Kokkos::complex<double>(3.));
+  static_assert(Kokkos::conj(1) == Kokkos::complex<double>(1));
+  static_assert(Kokkos::conj(2.f) == Kokkos::complex<float>(2.f));
+  static_assert(Kokkos::conj(3.) == Kokkos::complex<double>(3.));
 // long double has size 12 but Kokkos::complex requires 2*sizeof(T) to be a
 // power of two.
 #ifndef KOKKOS_IMPL_32BIT
-  ASSERT_EQ(Kokkos::conj(4.l), Kokkos::complex<long double>(4.l));
+  static_assert(Kokkos::conj(4.l) == Kokkos::complex<long double>(4.l));
   static_assert(
```
[the remark this answers is not in the document]
Comment (romintomasetti) [reply core/unit_test/TestComplex.hpp:471]: Indeed, I've added a test that exercises the constexpr-ness of the `conf` for a complex.

File: core/src/Kokkos_Complex.hpp [hunk 2]
```diff
@@ -789,9 +789,9 @@ KOKKOS_INLINE_FUNCTION Kokkos::complex<RealType> sqrt(
 
 //! Conjugate of a complex number.
 template <class RealType>
-KOKKOS_INLINE_FUNCTION complex<RealType> conj(
+KOKKOS_INLINE_FUNCTION constexpr complex<RealType> conj(
     const complex<RealType>& x) noexcept {
-  return complex<RealType>(real(x), -imag(x));
+  return {real(x), -imag(x)};
```
[the remark this answers is not in the document]
Comment (romintomasetti) [reply core/src/Kokkos_Complex.hpp:796]: Made them constexpr.

File: core/unit_test/TestComplex.hpp [hunk 3]
```diff
@@ -669,6 +668,12 @@ constexpr bool comparison_in_constant_expression() {
 
 static_assert(comparison_in_constant_expression());
 
+constexpr bool test_complex_conj() {
+  return Kokkos::conj(Kokkos::complex<double>{1., 1.}) ==
+         Kokkos::complex<double>{1., -1.};
+}
+static_assert(test_complex_conj());
+
```
Comment (tretre91) [inline core/unit_test/TestComplex.hpp:706]: Not blocking, but I would add tests for each of the possible combinations of positive and negative
```suggestion
constexpr bool test_complex_conj() {
  static_assert(Kokkos::conj(Kokkos::complex<double>{1., 1.}) ==
                Kokkos::complex<double>{1., -1.});
  static_assert(Kokkos::conj(Kokkos::complex<double>{1., -1.}) ==
                Kokkos::complex<double>{1., 1.});
  static_assert(Kokkos::conj(Kokkos::complex<double>{-1., 1.}) ==
                Kokkos::complex<double>{-1., -1.});
  static_assert(Kokkos::conj(Kokkos::complex<double>{-1., -1.}) ==
                Kokkos::complex<double>{-1., 1.});

  return true;
}
static_assert(test_complex_conj());

```

**Review by tretre91 — APPROVED** [review]
Looks good to me

File: core/unit_test/TestComplex.hpp (hunk 3 above)
Comment (romintomasetti) [reply core/unit_test/TestComplex.hpp:706]: OK, done.

**Review by tretre91 — APPROVED** [review]
Looks good to me

File: core/src/Kokkos_Complex.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (yasahi-hpc) [reply core/src/Kokkos_Complex.hpp:796]: It is out-of-the scoped of this PR, but should we make `real(const complex<T>& z)` and `imag(const complex<T>& z)` `constexpr` as well?
https://en.cppreference.com/w/cpp/numeric/complex/real2.html

**Review by yasahi-hpc — APPROVED** [review]
This PR looks good
Should we also make other free functions `constexpr` as well?

**Review by masterleinad — APPROVED** [review]

**Review by nliber — APPROVED** [review]
